### PR TITLE
🏗✨  Change the way the AMP canary runtime is built, tested (and released)

### LINKED ISSUE #17706: Ensure the ~1% canary is testing what gets rolled out to 100% during the release
After discussing with @erwinmombay and @rsimha, it appears that our current canary build runs with a [canary config](https://github.com/ampproject/amphtml/blob/master/build-system/global-configs/canary-config.json) rather than a [prod config](https://github.com/ampproject/amphtml/blob/master/build-system/global-configs/prod-config.json).  

The result is that we can't guarantee that what is being tested for ~1% of users in the canary is the same thing that gets rolled out to 100% of users.

For example an experiment may be turned on in canary but not turned on in prod.  In the current system the ~1% of users will have the experiment on, but when the release gets rolled out to 100% the experiment will be off.  Problems may not be exposed during the ~1% canary in this scenario--e.g. if the experiment enabled for the ~1% of users hides an issue in the release that is only present when the experiment is off.

We'd like to have a system that has these features:

- When we push a build that has been running at ~1% to 100% of users, the 100% should be the same binary and the same config that the ~1% was using.
- We want to maintain the ability to turn on experiments for a subset of users without requiring that the experiment is automatically turned on for all users as a result of the release process.

(Ideally config pushes would also be independent of the binary pushes, but that's beyond the scope of this issue.)

### PR DESCRIPTION
**Pull request by rsimha**
For the longest time, we have been testing the canary build of AMP with `build-system/global-configs/canary-config.json` and releasing it to production the following week with `build-system/global-configs/prod-config.json`. This is bad because the runtime gets tested with one set of config flags and released with another.

This PR fixes this testing hole by doing the following:
1. Consolidates all AMP config flags into one file `build-system/global-configs/amp-config.json`
    - Removes the `canary: 0` or `canary: 1` flag from the config file
2. Moves the logic that adds `canary: 0` or `canary: 1` flag to `AMP_CONFIG` to the build process
    - Removes the `--config` flag from `gulp build`, `gulp dist`, `gulp test`, and `gulp prepend-global`
    - Adds a `--canary` flag to `gulp build` and `gulp dist`. If omitted, `canary: 0` will be written to `AMP_CONFIG`. If used, `canary: 1` will be written to `AMP_CONFIG`.
4. Removes the (now-redundant) config-writing logic from `gulp test`, since this is taken care of during the build. (See related #18158 for `gulp visual-diff`.)
5. Introduces a new, unified means of specifying experiments, as described below.

This PR will need to be followed up by an update to our release scripts to use the new config filename.

**Here's how to add flags to `amp-config.json`:**

Use frequency values in the range `[0, 1]` for features that you would like to enable or disable. Examples:

- Use frequency value = 0 for features that are disabled for a 100% of users.
- Use frequency value = 1 for features visible to a 100% of users (production).
- Use frequency value = 0.01 for features visible to 1% of users (canary).
- If, for some reason, a feature needs to be visible to a fraction of canary
  users, use a fraction of 0.01. For example, 0.0001 == 1% of canary users.

Fixes #17706

### CODE REVIEW

**Comment by rsimha** [comment]
/to @cramforce @erwinmombay @danielrozenberg for a detailed review
/cc @choumx @jridgewell @aghassemi @lannka as an FYI, since this will change how we build, test, and release canary

**Comment by erwinmombay** [comment]
can you clarify how will the error reporter identify itself if its in canary or not with the canary field?

File: build-system/global-configs/amp-config.json [hunk 1]
```diff
@@ -15,39 +15,40 @@
     "twitter-default-placeholder-fade"
   ],
 
-  "canary": 0,
-  "expAdsenseA4A": 0.01,
+  "a4aFastFetchAdSenseLaunched": 0,
```
Comment (cramforce) [inline build-system/global-configs/amp-config.json]: Is the alpha ordering enforced now?

**Comment by rsimha** [comment]
Some responses (GH doesn't let you create conversations from top-level comments):

@erwinmombay Can you elaborate on how the error reporter works today? This PR doesn't remove the `canary` field from the `AMP_CONFIG` at the top of the runtime. All it does is change the mechanism for adding the field. Earlier, the field used to be copied directly from the JSON file. Now it's inserted by the build command, based on whether `--canary` was passed to it or not.

@cramforce This was a question I had too, since any new flag that is added (say, with a value of `0.01`) will automatically go to 1% of prod users the following week. I discussed this with @jridgewell, who said this shouldn't be a problem in practice, since the purpose of canary features is to have 1% of prod users see them.

One possible way to overcome this limitation might be to reserve the `-1` value for features that are on for opt-in canary users, but off for all users in prod. We'd need to add some code to the runtime to process and override this value.

Do you think a) this makes sense? b) is feasible? c) is a good idea?

File: gulpfile.js [hunk 2]
```diff
@@ -757,16 +757,15 @@ function printNobuildHelp() {
...
+function printCanaryHelp(command) {
   if (!process.env.TRAVIS) {
-    log(green('Building the runtime for local testing with the'),
-        cyan((argv.config === 'canary') ? 'canary' : 'prod'),
-        green('AMP config.'));
-    log(green('⤷ Use'), cyan('--config={canary|prod}'), green('with your'),
-        cyan(command), green('command to specify which config to apply.'));
+    log(green('Building the runtime for local testing in'),
+        cyan(argv.canary ? 'canary' : 'prod'), green('mode.'));
+    log(green('⤷ Use'), cyan('--canary'), green('with your'),
```
Comment (danielrozenberg) [inline gulpfile.js]: nit: wrap this line with `if (argv.canary)` - no need to tell the users to use `--canary` when they already used `--canary` 😉

File: build-system/tasks/prepend-global/index.js [hunk 3]
```diff
@@ -112,17 +112,17 @@ function valueOrDefault(value, defaultValue) {
 }
 
 /**
- * @param {string} config Prod or canary
+ * @param {boolean} canary Canary mode in AMP_CONFIG. Defaults to false.
  * @param {string} target File containing the AMP runtime (amp.js or v0.js)
- * @param {string} filename File containing the (prod or canary) config
+ * @param {string} filename File containing the AMP config
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/prepend-global/index.js]: Good point. I don't see a custom filename being used in any of our release scripts. @erwinmombay, can you think of any reason to change the source of the AMP config to be changed from the now single config file `build-system/global-configs/amp-config.json`?

File: gulpfile.js (hunk 2 above)
Comment (rsimha) [reply gulpfile.js]: Done, with `if (!argv.canary)`.

**Comment by cramforce** [comment]
Let's discuss this change in the next design review.

File: build-system/tasks/prepend-global/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (erwinmombay) [reply build-system/tasks/prepend-global/index.js]: hmm possibly other parties with their runtime ? i'd still like some flexibility in determining this during execution time

[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/prepend-global/index.js]: SGTM. Leaving this in.

**Comment by choumx** [comment]
One downside of using `0.01` for "pseudo-canary-only" is that error reports will appear in prod channel.

**Comment by rsimha** [comment]
Abandoning this approach (of merging config files) in favor of a new `rc` flavor release. See https://github.com/ampproject/amphtml/issues/17921#issuecomment-422954913

**Pull request closed without merging**